src/fparser/one/tests/test_do.py

```diff
@@ -0,0 +1,212 @@
+        assert str(loop).splitlines() == do_expected.splitlines()
+
+
+@pytest.mark.parametrize('name', [None, 'loop_name'])
+@pytest.mark.parametrize('label', [None, '123'])
+@pytest.mark.parametrize('control_comma', [False, True])
+@pytest.mark.parametrize('terminal_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('end_name', [None, 'loop_name', 'wrong_name'])
+@pytest.mark.parametrize('end_label', [None, '123', '456'])
+def test_do_while(name, label, control_comma, terminal_expression,
+                  end_name, end_label):
+    # pylint: disable=redefined-outer-name, too-many-arguments
+    '''
+    Checks that the "do while" loop parser understands all forms of the syntax.
+    '''
+    name_snippet = name + ': ' if name else None
+    label_snippet = label + ' ' if label else None
+    comma_snippet = ', ' if control_comma else None
+    code = '''{name}do {label}{comma}while ({term})
+  write (6, '(I0)') variable
+{endlabel} end do {endname}
+'''.format(name=name_snippet or '',
+           label=label_snippet or '',
+           comma=comma_snippet or '',
+           term=terminal_expression,
+           endlabel=end_label or '',
+           endname=end_name or '')
+    expected = '''  {name}DO {label}while ({term})
+    WRITE (6, '(I0)') variable
+{endlabel} END DO {endname}
+'''.format(name=name_snippet or '',
+           label=label_snippet or '',
+           term=terminal_expression,
+           endlabel=end_label or ' ',
+           endname=end_name or '')
+    print(code)
+    reader = FortranStringReader(code)
+    reader.set_format(FortranFormat(True, False))
+    parser = FortranParser(reader)
+    if (name != end_name) or (label and (label != end_label)):
+        with pytest.raises(AnalyzeError):
+            parser.parse()
+    else:
+        parser.parse()
+        loop = parser.block.content[0]
+        assert str(loop).splitlines() == expected.splitlines()
+
+
+@pytest.mark.parametrize('name', [None, 'loop_name'])
+@pytest.mark.parametrize('label', [None, '123'])
+@pytest.mark.parametrize('control_comma', [False, True])
+@pytest.mark.parametrize('typespec',
+                         [None, 'logical', 'real(r16)', 'integer(kind=i32)'])
+@pytest.mark.parametrize('initial_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('terminal_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('incrament_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('mask_expression',
+                         ['1', '10', 'x+y', 'size(array)', 'size(this%array)'])
+@pytest.mark.parametrize('end_name', [None, 'loop_name', 'wrong_name'])
+@pytest.mark.parametrize('end_label', [None, '123', '456'])
+def test_do_concurrent(name, label, control_comma, typespec,
```

Comment:
Is 'do concurrent' F2008? If so, please document this in the fparser1 user guide.

Reply:
It is and as such I have removed it from the testing. This seemed appropriate as there is no guarantee that fparser-1 will support anything beyond Fortran 95. As it happens it will support this because it treats everything after "do" as a plain string but since you are keen to cut down the amount of testing...